Add Negotiations mission (Red Alert, scg47ea)

JovialFeline opened the pull request:
This adds Negotiations from Aftermath. It's a short mission featuring Tanya, some hostages in need of rescue, and many barrels in need of exploding.

Normal's more of a straight conversion. Thoughts on the Hard changes are welcome.

Feedback on the cameras would be dandy. The original relies on vision of old areas to avoid patrols, so cameras are used to give information when needed.
 - Standard reveals are the usual 6s, while more event-y ones are 20s.
 - Most barrel explosions also come with a 1s camera reveal, for fun and confirming kills.

----

### Design changes:

**Original:** A hostage must die before the group will follow Tanya.
**New:** All hostages can be saved.

**Original:** If he lives, the helpful Guide will either follow Tanya until the mission ends, or stand still near the hostage fencing.
**New:** He can be escorted back to the church like the other civilians.

**Original:** Soviets produce tanks as soon as a War Factory is up, including Tesla Tanks.
**New:** Soviets will build a Service Depot before producing tanks. They also start with a Soviet Tech Center, which can be destroyed to stop production of Tesla Tanks or Shock Troopers.

**Original:** The General will retreat to the northwestern beach when Tanya or the Guide come within ~7 cells. A Soviet MCV will arrive if the General's left alone.
**New:** The General is faster and the retreat is activated sooner. Reinforcements will come if the General stays for ~~45s~~ -> 30s.
 - This is done because of ORA Tanya's better speed and range, and because of the added tank delay from the Service Depot.
 - The reinforcement timing change doesn't seem drastic. The MCV boat should arrive as Tanya and the hostages are ~~entering the forest~~ -> near the forest patrol's starting point.

The MCV's arrival is revealed, and there's a notification for the General's flare.
 
Swapped some unit types around for convention's sake.
 - Soviet Chinook reinforcements now enter from a Badger.
 - Aside from Tanya, Allied Badger reinforcements now enter from Chinooks.
 - The 3 Rocket Soldiers owned by the Soviets are swapped for Flamethrowers.

The Allied reinforcements come with 2 more mechanics.
 - May need adjusting. The attrition can be rough with a solo mechanic if Soviet reinforcements are signaled. The original solo mechanic is bonkers, healing a medium tank from near death in ~10s compared to ~64s in ORA.

----

### Hard difficulty:

All hostages must survive.

Tanya will encounter more early resistance.

 - If Tanya dies from a rocket, the launcher will be revealed (again) for the sake of clarity.

The return trip to the starting beach has some added enemies.

There is some more security around the prison and the BadGuy base area (where the MCV can deploy).

It's possible for Soviets to build Shock Troopers.

----

### Notes on porting and such:

To mimic [their original guard range](https://github.com/OpenRA/OpenRA/assets/4985264/fa691780-42da-4445-9400-b6d1d53d7103), some dogs have a wider AutoTarget ScanRadius.
 - I felt less kooky once I saw that allies-03a/allies-03b also did this for its dogs.
 - While the original rules.ini says `GuardRange=7`, 9 seemed to be a closer match.

Guide's house is more vulnerable to damage, since ORA rifles do little to structures.
 - The house will take ~11.56s to die if left alone. Original is ~10s.
 - Could have reduced the HP 53% -> 23%, but wanted to preserve the crumbling structure effect.
 
This western passage is originally passable, so I cleared a path. The eastern passage has a (narrow) path that works, so it was left untouched.
![RA95_RubbleWest](https://github.com/OpenRA/OpenRA/assets/4985264/941a5042-cab7-403a-baf6-795d95911aca)

The Demolition Truck's explosion range is tightened, closer to the original's.
 - The Soviet truck can otherwise explode barrels by the prison, freeing the hostages.

The Chrono Tank reinforcements' entry and triggers are slightly moved to avoid spawning inside other units.

- I don't know of a current way to order the Chrono Tanks' PortableChrono. I've instead implied that a Chronosphere is used to transport them.
  - This is consistent with Production Disruption or In The Nick of Time anyway.
  - They will be fully charged, and the Chronosphere sound is played instead of the tank shift sound.
  - I did manage the palette shift by moving a dummy unit. Mousetrap did something similar.

The patrolling tank encountered during the rescue is guarded by two Shock Troopers, so it's fortunately still a threat to ORA Tanya.

The Guide's movement is more direct.
- He no longer pauses at the southern river crossing since ORA Tanya has no issue keeping up.
- After the forest patrol is clear, he will head straight for the village bridge, instead of moving to the forest's middle or the forest patrol's starting point.

In case you're left to wonder why the Soviets' Demolition Truck goes hunting, that is the original `hunu` trigger.
 - BadGuys' units will hunt if Allied reinforcements have arrived and the General is dead.
 - This does not include units set to Area Guard. That leaves the truck, a handful of Flamethrowers, and BlockerTeam (1 Heavy Tank with 2 Shock Troopers).
 
For funsies, hostages following Tanya will panic if she dies. Their disregard for personal space in RA '95 meant they were likely to do it there as well.

PunkPun commented:
> Swapped some unit types around for convention's sake.
>
> * Soviet Chinook reinforcements now enter from a Badger.
>* Aside from Tanya, Allied Badger reinforcements now enter from Chinooks.

I think this can be reversed. If anything we would like to give chinook to both factions like in vanilla

PunkPun commented:
I think the flare next to the general can extinguish after he's been killed, maybe a few minutes death timer? It feels a bit off when it's permanent.

Also the camera on the demo truck should be triggered from the other side of the fence. I've also noticed that if you don't blow up the demo while on the other side of the fence, it's a big struggle to keep the civilians alive

I'm also a bit stumped, if you get into this situation how do you proceed?

![grafik](https://github.com/OpenRA/OpenRA/assets/37534529/768c27f7-a444-4480-92dc-b6e4e28b9fc7)

PunkPun commented on mods/ra/languages/lua/en.ftl:
```diff
@@ -274,7 +288,7 @@ destroy-soviet-convoy = Destroy all Soviet convoy trucks.
     convoys down.
 
-## siberian-conflict-3-wasteland
+## siberian-conflict-3-wasteland, negotiations
```
should also mention the mission for the `tanya-rules-of-engagement` translation

PunkPun reviewed (commented):
ah I understand, but it's almost impossible to win as civilians will get themselves killed very quickly onto the infantry and flame tower

JovialFeline commented:
> I think the flare next to the general can extinguish after he's been killed, maybe a few minutes death timer? It feels a bit off when it's permanent.

How long should that be? I know from fiddling with Don't Drink The Water that the original flares seem to expire after 8m but I'm guessing that'd be too long.

> Also the camera on the demo truck should be triggered from the other side of the fence. I've also noticed that if you don't blow up the demo while on the other side of the fence, it's a big struggle to keep the civilians alive

It can be, though in that case, you should have barrels for the village patrol; they move counter-clockwise and the dog especially should be easy to burn.

The demo truck camera should already spawn at the truck's death location unless it's attacking or I've goofed and this trigger only does it at its starting location.
```lua
	Trigger.OnKilled(SovietDemoTruck, function()
		if not SovietDemoTruck.IsIdle then
			return
		end
		SpawnPlayerCamera(SovietDemoTruck.Location, DateTime.Seconds(1), "camera.paradrop")
	end)
```
> I'm also a bit stumped, if you get into this situation how do you proceed?

The intended route seems to be returning to the forest. Once there, the tank and shock troopers at the river will move through, north toward the village bridge. You can use the same hideout on the east edge to wait them out, then proceed to the church.

I'm a little irked to see them right at the river crossing instead of in the middle of the open church area, though. Odd.

JovialFeline commented:
> I think the flare next to the general can extinguish after he's been killed, maybe a few minutes death timer? It feels a bit off when it's permanent.

How long should that be? I know from fiddling with Don't Drink The Water that the original flares seem to expire after 8m but I'm guessing that'd be too long.

> Also the camera on the demo truck should be triggered from the other side of the fence. I've also noticed that if you don't blow up the demo while on the other side of the fence, it's a big struggle to keep the civilians alive

It can be, though in that case, you probably have barrels for the village patrol. They move clockwise and the dog especially should be easy to burn.

The demo truck camera _should_ already spawn at the truck's death location unless it's attacking or I've goofed and this trigger only does it at its starting location.
```lua
	Trigger.OnKilled(SovietDemoTruck, function()
		if not SovietDemoTruck.IsIdle then
			return
		end
		SpawnPlayerCamera(SovietDemoTruck.Location, DateTime.Seconds(1), "camera.paradrop")
	end)
```
> I'm also a bit stumped, if you get into this situation how do you proceed?

The intended route seems to be returning to the forest. Once there, the tank and shock troopers at the river will move through, north toward the village bridge. You can use the same hideout on the east edge to wait them out, then proceed to the church.

I'm a little irked to see them idle at the river crossing instead of the open church area, though. Odd.

PunkPun commented:
> The demo truck camera should already spawn at the truck's death location

Yeah it spawns, I mean the camera revealing the barrels

JovialFeline commented:
### Update

The tank-guarding Shock Troopers are now Flamethrowers to reduce insta-deaths from fog.
  - At least for now, the forest team (with dogs) and northern Chinook team still have their Shock Troopers.

Allies are back to using Badgers for all their infantry drops.

Soviets use Chinooks and a handful of rocket launchers again.
  - There's some randomness in what northern edge cell the original Chinook spawns, but 9s until landing seems to be the average.

The explosion reveals should feel more responsive.

The money check for bot structure building is improved.

The General's flare will expire 30 seconds after he does. Tanya's flare expires after 5 minutes.

The timer now changes with difficulty.
  - Easy: 5 minutes
  - Normal: 4 minutes
  - Hard: 3 minutes
  - "every 5 minutes" is now "every few minutes" in the briefing text.

The "Keep all hostages alive" objective is still required for Hard, but also appears on Normal or Easy as an optional objective.

`negotiations` is listed with Tanya's usual targeting translation.

The copyright notice is added.

PunkPun commented:
It's definitely easier with less shokies and with working patrols. Just the tiny cameras felt like there were enabled for too short. It's a blink it and you'll miss it

Without reading the briefing it might be difficult to to tell from where to call in allied reinforcements. I'm not wether we need to fix it, but it's something worth thinking about

PunkPun commented:
Without reading the briefing it might be difficult to to tell from where to call in allied reinforcements. I'm not wether we need to fix it, but it's something worth thinking about

JovialFeline commented:
Small changes:
- Objectives now use the common utility functions, which had slipped my mind.
- After hostages make it inside the church, [a looping beacon is created](https://i.imgur.com/RrAGBfr.gifv) at the starting beach. It will stop looping after Tanya triggers reinforcements.

PunkPun reviewed (commented):
Perhaps create the "Signal for reinforcements" objective when it's actually possible to do. Instead of the ping the objective could just specify the location

Could delay the church objective as well, so there's less to read at the start of the mission. Ofc mission should still fail on blowing up the church.

When the timer ran out I got this crash
`Actor 'c2 (dead)' does not define a property 'Kill'`

JovialFeline commented:
Thank you. I can't figure how that re-kill may have happened without Tanya shooting the hostage, but that possibility was easy enough to test and the new check should handle any premature deaths.

---

- If a hostage scheduled for execution is already dead, a random survivor will be chosen instead. If every hostage is dead, execution is skipped.

- The escort and signal objectives are delayed until they're more relevant. The beach beacon is removed.

- The destroy objective is declared after the hostages are escorted. It's still possible to destroy all Soviets for victory without reinforcements, so long as hostages are escorted to the church at some point.

- A Chrono Tank timing goof is fixed. The tanks can be triggered after ~~hostages enter the church~~ -> Allied reinforcements are signaled.

- The hostage pen's executioner is set to Defend stance.

- The returning forest patrol will have another waypoint in their path so their dogs won't run ahead as much, and another reveal is added. The reveals for that group's return can trigger after ~~hostages are freed~~ -> after the group moves all the way south to the Guide house, like Blocker Team does for their own reveal.

- Blocker Team's path to the Guide house skips the forest patrol's starting point and makes a traffic jam less likely.

PunkPun reviewed (commented):
Should the tech centre be sold off upon the destruction of the command centre? Because it isn't

The mission is better now that there's less objective at the start, but you are still bombarded with text when you return the hostages to the church. I dunno how but it should be best to spread it out a bit more.

Signal flare detected could also be more specific. Perhaps flare detected in the west?

The church itself could also be highlighted somehow? I think it would be best to somehow highlight it upon saving the helpful civilian

I also really enjoyed the extra troops on hard difficulty. They made the mission feel more complete

JovialFeline commented:
- I've considered including that Tech Center in the fire sale if the General never signaled for help. It's owned by BadGuy (owner of the MCV). It should be easy enough to add once that check is done; `FireSale` does its thing with whatever player gets passed as an argument.

- I'll have to ponder the objective stuff more. Ideally, I'd add the destroy objective after a delay but the game will see all objectives done and trigger victory. That is less of a problem on Hard where saving every hostage is required and only gets ticked done once all Soviets are gone.

- Funny enough, I used to have SignalFlareNorth as the notification but went back and forth on it. Perhaps the west one will fit better with how the player's map is gonna look.

- D'ya have any particular ideas for the church? Should I bring back the beacon?  :p

- Normal indeed feels a bit empty to me when I play it now, yeah. Glad you dig it.

PunkPun commented:
perhaps change the objective to 
```diff
-Get hostages to church.
+Get hostages to town.
```
and once you cross the river make the civilians go to church automatically?

The helpful civilian already calls this place town

JovialFeline commented:
Took longer than expected, but some more changes got done.

----

Made some vision tweaks so the church should be more obvious, and the church itself will provide vision until a hostage is loaded. The trigger ordering hostages to enter is extended to match that vision circle.

Changed objectives so their messages are spaced out a bit, even the primary ones.

Swapped the SignalFlare notification for SignalFlareWest.

Two oddball crash scenarios are fixed.
  - Tanya bombs Guide's house shortly before he's added to the world.
  - BadGuy attempts to build a structure with an owned husk in the way.

The Guide will reroute if the bridge is out.
  - He will also have a quicker reaction to the forest patrol's death and the hostages' freedom.

The northern edge team is moved a bit so they won't be partly obscured by the edge shroud.

Blocker Team will correctly start moving south as the prison is revealed, rather than when the General retreats.

BadGuy's units near the forward base will join in USSR's attack if it's destroyed.

The Soviets can sell off their Tech Center when appropriate to speed up the end.
<details><summary>Details:</summary>
It will be sold after one of two events: the Forward Command is destroyed or the base built by BadGuy's reinforced MCV becomes useless. It's considered useless when that base is missing its Construction Yard, defenses, War Factory, and Barracks. Timelines I've considered and tested:

  - A: The Forward Command is destroyed, triggering USSR's fire sale. The General did not signal for help, so BadGuy's MCV will never come.
    - The Tech Center will be included in USSR's fire sale.

  - B: The Forward Command is destroyed after BadGuy's MCV arrival. The Tech Center remains intact.
    - If BadGuy's base is still okay, don't sell the Tech Center yet.
    - If BadGuy's base becomes useless, sell the Tech Center and send its occupants hunting.
      - If the difficulty is Hard, the remainder of BadGuy's base will do the same.

  - C: The Forward Command is intact after BadGuy's MCV arrival.
    - If BadGuy's base becomes useless, USSR will take ownership of the Tech Center. It will later be sold as part of the USSR fire sale.

  - D: The Forward Command is intact after BadGuy's MCV arrival, but Tanya demolishes the MCV before it can deploy.
    - The result will be like a useless base in timeline B.
</details>

PunkPun reviewed (dismissed):
The mission plays really well, LGTM

abcdefg30 commented on mods/ra/maps/negotiations/map.yaml:
```diff
@@ -0,0 +1,1351 @@
+		Faction: soviet
+		Color: FF8800
+		Allies: USSR
+		Enemies: Greece, England
```
```suggestion
		Enemies: Greece, England, GoodGuy
```

?

abcdefg30 replied:
(Seems like all the Allies/Enemies setups are a bit off.)

JovialFeline replied:
That's indeed a mistake, though the two shouldn't come into contact. England and Germany are intentionally a bit skewed, but all the diplomacy could use some changes. Thanks.

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-ai.lua:
```diff
@@ -0,0 +1,349 @@
+	)
+
+	Trigger.OnEnteredFootprint( { BuilderRally.Location }, function(actor, id)
+		if not (actor.Type == "fact") then
```
```suggestion
	Trigger.OnEnteredFootprint({ BuilderRally.Location }, function(actor, id)
		if actor.Type ~= "fact" then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-ai.lua:
```diff
@@ -0,0 +1,349 @@
+IsNewBaseOkay = function()
+	local types = { "fact", "barr", "ftur", "weap", "tsla" }
+
+	local okay = Utils.Any(types, function(type)
+		return BadGuy.HasPrerequisites( { type } )
+	end)
+
+	return okay
```
```suggestion
	return Utils.Any(types, function(type)
		return BadGuy.HasPrerequisites({ type })
	end)
```

JovialFeline replied:
Oops. That was left behind when I removed the many, _many_ debug messages.

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-ai.lua:
```diff
@@ -0,0 +1,349 @@
+end
+
+IsHarvesterNeeded = function(player)
+	return player.HasPrerequisites( { "proc" } ) and #player.GetActorsByType("harv") < 1
```
```suggestion
	return player.HasPrerequisites({ "proc" }) and #player.GetActorsByType("harv") < 1
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-ai.lua:
```diff
@@ -0,0 +1,349 @@
+		return false, "blocked"
+	end
+
+	local actor = Actor.Create(structure.type, true, { Owner = player, Location = structure.location } )
```
```suggestion
	local actor = Actor.Create(structure.type, true, { Owner = player, Location = structure.location })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+
+LinkCooperativeObjectives = function()
+	local greeceActor = Greece.GetActorsByType("player")[1]
+	Trigger.Clear(greeceActor, "OnPlayerWon")
```
Why is this needed?

JovialFeline replied:
It has to do with the way objectives are set up to address https://github.com/OpenRA/OpenRA/pull/20975#pullrequestreview-1608535624 's bit about objective message spam. The mission has points where this can happen:

1. Greece completes a primary objective.
2. Greece has completed all primary objectives, and is considered a winner. This triggers the OnPlayerWon added by InitObjectives from `campaign.lua`.
3. Mission completion isn't triggered because this is a co-op mission under the hood, and their ally England needs to also win.
4. After a delay, Greece receives a new primary objective and the mission carries on.

At step 2, there would normally be the speech announcement "Mission accomplished." The options I saw were to remove the trigger from Greece, or put a rewritten InitObjectives into the mission. I figured the first option was less awkward. A little after this line, England's win hands Greece the normal announcement.

Ideally, I'd just set primary objectives at the start and delay showing/announcing them in a more straightforward way, but I realize that's a bit niche.

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+	Media.PlaySoundNotification(Greece, "Chronoshift")
+
+	Utils.Do(locations, function(location)
+		Actor.Create("ctnk", true, { Owner = Greece, Location = location, Facing = Angle.NorthWest } )
```
```suggestion
		Actor.Create("ctnk", true, { Owner = Greece, Location = location, Facing = Angle.NorthWest })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+	-- For the palette shift that comes with the tanks' arrival.
+	local units = { }
+	units[ChronoDummy] = ChronoDummy.Location + CVec.New(1, 0)
+	proxy.Chronoshift(units, 0)
```
Why not directly chronoshift the actual units? See how the desert shellmap does it:
```
ChronoshiftAlliedUnits = function()
	local cells = Utils.ExpandFootprint({ ChronoshiftLocation.Location }, false)
	local units = { }
	for i = 1, #cells do
		local unit = Actor.Create("2tnk", true, { Owner = Allies, Facing = Angle.North })
		BindActorTriggers(unit)
		units[unit] = cells[i]
	end
	Chronosphere.Chronoshift(units)
	Trigger.AfterDelay(DateTime.Seconds(60), ChronoshiftAlliedUnits)
end
```

JovialFeline replied:
For whatever reason, I've had problems getting units to shift to the correct location unless they're shifted more than once, and possibly more depending on distance between the start and goal. This doesn't seem to happen if both locations are tucked into a corner. I don't know the cause yet and I'll have to do some more experimenting.

JovialFeline replied:
Scratch my previous comment for now. It seems whatever was going on with Actor.Teleport (and chronoshifts by extension) is fixed, or that's what digging up my old test map indicates, anyway.

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+
+	Trigger.OnAddedToWorld(plane, function()
+		Utils.Do(passengers, function(type)
+			local passenger = Actor.Create(type, false, { Owner = Greece } )
```
```suggestion
			local passenger = Actor.Create(type, false, { Owner = Greece })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+
+	local waterCamera = SpawnPlayerCamera(NorthWaterEntry.Location, -1)
+	local boatPath = { NorthWaterEntry.Location, BuilderBeach.Location }
+	local cargo = Reinforcements.ReinforceWithTransport(BadGuy, "lst", { "mcv" }, boatPath, { boatPath[1] } )[2]
```
```suggestion
	local cargo = Reinforcements.ReinforceWithTransport(BadGuy, "lst", { "mcv" }, boatPath, { boatPath[1] })[2]
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+ReinforceNorthChinook = function()
+	local delay = DateTime.Seconds(4)
+	local camera = SpawnPlayerCamera(NorthChinookUnload.Location, -1, "camera.small", delay + DateTime.Seconds(2))
+	local transport = Actor.Create("tran.north", false, { Owner = USSR, Location = NorthChinookEntry.Location, Facing = Angle.South } )
```
```suggestion
	local transport = Actor.Create("tran.north", false, { Owner = USSR, Location = NorthChinookEntry.Location, Facing = Angle.South })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+	local passengers = transport.Passengers
+
+	Trigger.AfterDelay(delay, function()
+		transport.IsInWorld = true
```
This is kinda weird, why not
```
Trigger.AfterDelay(delay, function()
	local transport = Actor.Create("tran.north", true, { Owner = USSR, Location = NorthChinookEntry.Location, Facing = Angle.South } )
	transport.UnloadPassengers(NorthChinookUnload.Location)
	transport.Move(SouthChinookExit.Location)
	transport.Destroy()

	Trigger.OnPassengerExited(transport, function()
		if transport.PassengerCount > 0 then
			return
		end

		-- Original unload is slow. Mimic the timing with a short pause.
		Trigger.AfterDelay(DateTime.Seconds(3), function()
			OrderNorthChinookSoldiers(transport.Passengers, camera)
		end)
	end)

	Trigger.OnAnyKilled(transport.Passengers, function()
		RemoveActor(camera, DateTime.Seconds(1))
	end)
end)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+
+		Trigger.AfterDelay(i * 5, function()
+			guard.IsInWorld = true
+		end)
```
Same here:
```suggestion
		Trigger.AfterDelay(i * 5, function()
			local guard = Actor.Create("e1", false, { Owner = USSR, Location = GuardHouseSpawn.Location, SubCell = 4, Facing = Angle.SouthWest })
			guard.Move(GuardHouseExit.Location)
			guard.AttackMove(goal)
			IdleHunt(guard)
		end)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+end
+
+ReinforceHardTeams = function()
+	if not (Difficulty == "hard") then
```
```suggestion
	if Difficulty ~= "hard" then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+		actor.Scatter()
+	end)
+
+	Trigger.OnEnteredFootprint( { GeneralRally.Location }, function(actor, id)
```
```suggestion
	Trigger.OnEnteredFootprint({ GeneralRally.Location }, function(actor, id)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+	end)
+
+	Trigger.OnEnteredFootprint( { GeneralRally.Location }, function(actor, id)
+		if not (actor.Type == "gnrl") then
```
```suggestion
		if actor.Type ~= "gnrl" then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+end
+
+ReinforceSamRocket = function()
+	local rocket = Actor.Create("v2rl", true, { Owner = USSR, Facing = Angle.West, Location = SamRocketEntry.Location } )
```
```suggestion
	local rocket = Actor.Create("v2rl", true, { Owner = USSR, Facing = Angle.West, Location = SamRocketEntry.Location })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+
+PrepareSouthChinook = function()
+	local startPosition = SouthChinookEntry.CenterPosition + WVec.New(0, 0, Actor.CruiseAltitude("tran"))
+	local transport = Actor.Create("tran.south", false, { Owner = USSR, CenterPosition = startPosition, Facing = Angle.East } )
```
```suggestion
	local transport = Actor.Create("tran.south", false, { Owner = USSR, CenterPosition = startPosition, Facing = Angle.East })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+		end
+
+		Trigger.RemoveProximityTrigger(id)
+		transport.IsInWorld = true
```
Again, can we `Actor.Create` the unit here instead of adding things manually to the world?

JovialFeline replied:
Will do for this and the others. For this one, I believe I was trying to get the creation done and out of the way by the mission's start instead of doing it mid-game. I think it's both pointless and less readable as-is though.

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+end
+
+ReinforceHardDogs = function()
+	local startDog = CreateHardDog(WestRoadEntry.Location, Angle.South, { TanyaDrop.Location } )
```
```suggestion
	local startDog = CreateHardDog(WestRoadEntry.Location, Angle.South, { TanyaDrop.Location })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations-reinforcements.lua:
```diff
@@ -0,0 +1,409 @@
+end
+
+CreateHardDog = function(location, facing, triggerCells)
+	local dog = Actor.Create("dog.long", false, { Owner = USSR, Location = location, Facing = facing or Angle.South } )
```
```suggestion
	local dog = Actor.Create("dog.long", false, { Owner = USSR, Location = location, Facing = facing or Angle.South })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	Trigger.OnAllKilled(guideHouseAttackers, OnGuideRiflesKilled)
+
+	Trigger.OnEnteredProximityTrigger(TanyaDrop.CenterPosition, WDist.FromCells(4), function(actor, id)
+		if not (EscortFinished and actor.Type == TanyaType) then
```
```suggestion
		if not EscortFinished or actor.Type ~= TanyaType then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+			Trigger.RemoveProximityTrigger(id)
+			return
+		end
+		if not (actor.Owner == Neutral) then
```
```suggestion
		if actor.Owner ~= Neutral then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+		return
+	end
+
+	local text = UserInterface.Translate("hostage-dies-in", { ["time"] = Utils.FormatTime(TimeRemaining) } )
```
```suggestion
	local text = UserInterface.Translate("hostage-dies-in", { ["time"] = Utils.FormatTime(TimeRemaining) })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+end
+
+PrepareCountdown = function(interval)
+	local text = UserInterface.Translate("hostage-dies-in", { ["time"] = Utils.FormatTime(interval) } )
```
```suggestion
	local text = UserInterface.Translate("hostage-dies-in", { ["time"] = Utils.FormatTime(interval) })
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	local patrolPassed = false
+
+	local forestFoot = Trigger.OnEnteredFootprint( { ForestSafetyCheck.Location }, function(actor, id)
+		if patrolPassed or not (actor.Owner == USSR) then
```
```suggestion
		if patrolPassed or actor.Owner ~= USSR then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	ForestTeamReachedHouse = false
+	local patrolPassed = false
+
+	local forestFoot = Trigger.OnEnteredFootprint( { ForestSafetyCheck.Location }, function(actor, id)
```
```suggestion
	local forestFoot = Trigger.OnEnteredFootprint({ ForestSafetyCheck.Location }, function(actor, id)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+		GuideToVillage(nil, true)
+	end)
+
+	Trigger.OnEnteredFootprint( { ForestCenter.Location }, function(actor, id)
```
```suggestion
	Trigger.OnEnteredFootprint({ ForestCenter.Location }, function(actor, id)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+		end)
+	end)
+
+	Trigger.OnExitedFootprint( { ForestHideout.Location }, function(actor, id)
```
```suggestion
	Trigger.OnExitedFootprint({ ForestHideout.Location }, function(actor, id)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	end)
+
+	Trigger.OnExitedFootprint( { ForestHideout.Location }, function(actor, id)
+		if Executioner.IsDead or not (IsAlive(actor) and IsGuide(actor)) then
```
```suggestion
		if Executioner.IsDead or not IsAlive(actor) or not IsGuide(actor) then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+		Media.DisplayMessage(UserInterface.Translate("guide-follow-me"), actor.TooltipName)
+	end)
+
+	Trigger.OnEnteredFootprint( { ForestSouthwest.Location }, function(actor, id)
+		if not (actor.Type == "dog") then
```
```suggestion
	Trigger.OnEnteredFootprint({ ForestSouthwest.Location }, function(actor, id)
		if actor.Type ~= "dog" then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+		end)
+	end)
+
+	Trigger.OnEnteredFootprint( { GuideHouseReveal.Location }, function(actor, id)
+		if not (actor.Type == "dog") then
```
```suggestion
	Trigger.OnEnteredFootprint({ GuideHouseReveal.Location }, function(actor, id)
		if actor.Type ~= "dog" then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	Utils.Do(sides, function(side)
+		Trigger.OnEnteredFootprint(side.cells, function(actor, id)
+			local signaled = SignalForReinforcements and Greece.IsObjectiveCompleted(SignalForReinforcements)
+			if not (signaled and actor.Owner == Greece) then
```
```suggestion
			if not signaled or actor.Owner ~= Greece then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+PrepareSovietGuards = function()
+	Utils.Do(BadGuy.GetGroundAttackers(), function(soviet)
+		Trigger.OnDamaged(soviet, function(_, attacker)
+			if attacker.Type == "dtrk" or not (attacker.Owner == Greece) then
```
```suggestion
			if attacker.Type == "dtrk" or attacker.Owner ~= Greece then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	local houseReached = false
+
+	Trigger.OnEnteredProximityTrigger(BlockerReturnProximity.CenterPosition, WDist.FromCells(2), function(actor, id)
+		if not (Executioner.IsDead and actor.Owner == Greece) then
```
```suggestion
		if not Executioner.IsDead or actor.Owner ~= Greece then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+			return
+		end
+
+		if not (Executioner.IsDead and actor == BlockerTank) then
```
```suggestion
		if not Executioner.IsDead or actor ~= BlockerTank then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+			return
+		end
+
+		if not (houseReached and Executioner.IsDead and actor == BlockerTank) then
```
```suggestion
		if not houseReached or not Executioner.IsDead or actor ~= BlockerTank then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	PrepareMainReveals()
+
+	Trigger.OnEnteredProximityTrigger(Prison.CenterPosition, WDist.FromCells(10), function(actor, id)
+		if not (IsGuideOrAllies(actor)) then
```
```suggestion
		if not IsGuideOrAllies(actor) then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+
+	Utils.Do(actors, function(actor)
+		Trigger.OnDamaged(actor, function(_, attacker)
+			if revealed or not (attacker.Type == "barl" or "brl3") then
```
```suggestion
			if revealed or (attacker.Type ~= "barl" and attacker.Type ~= "brl3") then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+	local alerted = false
+	Utils.Do(actors, function(actor)
+		Trigger.OnDamaged(actor, function(_, attacker)
+			if alerted or attacker.Type == "dtrk" or not (attacker.Owner == targetPlayer) then
```
```suggestion
			if alerted or attacker.Type == "dtrk" or attacker.Owner ~= targetPlayer then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+				end
+				-- Patrolling groups will need to halt.
+				hunter.Stop()
+				hunter.Hunt()
```
`IdleHunt`?

JovialFeline replied:
Some of these will be groups using OnIdle to form up on a patrol location, wait until all live members of the group are there, then form up on the next location. An OnIdle clear followed by IdleHunt might be worth doing, though.

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+end
+
+SpawnGuide = function()
+	local guide = Actor.Create("c1", false, { Owner = GoodGuy, Location = GuideHouse.Location, SubCell = 4, Facing = Angle.South } )
+
+	Trigger.OnAddedToWorld(guide, function()
+		GuideToForest(guide)
+	end)
+
+	Trigger.AfterDelay(15, function()
+		if GuideHouse.IsDead then
+			return
+		end
+		guide.IsInWorld = true
+	end)
```
```suggestion
	Trigger.AfterDelay(15, function()
		if GuideHouse.IsDead then
			return
		end

		local guide = Actor.Create("c1", true, { Owner = GoodGuy, Location = GuideHouse.Location, SubCell = 4, Facing = Angle.South })
		GuideToForest(guide)
	end)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+			return
+		end
+
+		if not (actor.Type == "e1" and guide.Location == GuideBarrelGoal.Location) then
```
```suggestion
		if actor.Type ~= "e1" or guide.Location ~= GuideBarrelGoal.Location then
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+
+	guide.Move(GuideBarrelGoal.Location)
+	local revoked = false
+	local revokeTargets = Utils.Where( { Executioner, PrisonBarrel1, PrisonBarrel2, PrisonBarrel3, PrisonBarrel4 }, IsAlive)
```
```suggestion
	local revokeTargets = Utils.Where({ Executioner, PrisonBarrel1, PrisonBarrel2, PrisonBarrel3, PrisonBarrel4 }, IsAlive)
```

abcdefg30 commented on mods/ra/maps/negotiations/negotiations.lua:
```diff
@@ -0,0 +1,1237 @@
+
+RemoveActor = function(actor, delay)
+	Trigger.AfterDelay(delay or 0, function()
+		if not (actor and actor.IsInWorld) then
```
```suggestion
		if not actor or not actor.IsInWorld then
```

abcdefg30 reviewed (commented):
Looks like sublime work! Left a few comments which are mostly style nits.

JovialFeline commented:
The last time I made changes, this got stalled for a while with a crash until https://github.com/OpenRA/OpenRA/pull/21076 fixed that. This is rebased now with other minor adjustments after rereading and retesting.

### Changes from September:

(Where did the year go?)

Changed Chrono Tanks to shift into the fight themselves. The dummy unit and its rules have been removed.

Simplified some soldier reinforcements, along with the Guide's spawn.

<details><summary>Simplified Hard dog reinforcements.</summary>
There was some unnecessary code left from experiments with dogs in the village. Some dogs were created but not in the world until triggered, to avoid their early death from the demo truck explosion. Both dogs are now created normally at the start and given later orders with triggers.
</details>

Style tweaks galore.

### Changes from today:
Pointed map.yaml at the new(ish) `campaign/en.ftl` and changed the prison's tooltip to use its translation.

Added a death check to the forest patrol's dog whine.

Added some comments and renamed some things for clarity's sake.

The Guide is extra forbidden from shooting barrels near the prison if the hostages are already free. He still had a chance to do this under certain circumstances:
  - Tanya killed the executioner directly, leaving behind barrels.
  - Tanya lingered north of the prison as Soviet infantry approached the intact barrels.
  - The Guide stood at Waypoint 22, his favorite barrel shooting spot.
  
The Guide is no longer too polite to infiltrate a friendly church.

----

**Dec 19:** BadGuy's pause between structure (re)builds is a few seconds shorter.

----

**Feb 10:** Rebased.

----

**Feb 24**

Made the StartGameNotification change more graceful.
  - Monster Tank Madness and Mousetrap should probably get this treatment too so load/save speeches play.

Fixed a pair of goofs where I forgot `OnAnyKilled` triggers will only fire once.

Added a death check to `GroupHuntOnDamaged` just in case.

Fixed potential crashes for two unlikely events:
  - The Guide dies the instant before he stops to detour around a broken bridge.
  - On Hard, Tanya takes the north route past the tank, kills one of the scattered beach guards with explosive barrels, and does this before the General can arrive there.

Added and tweaked some more comments.

Porenutak commented:
On hard difficulty the first dog appear while tanya is still in air. On faster speeds its hard micro. Eats me little 10x. I would advice to delay dog at least for 1 sec. Otherwise looks good. Tested on all diffculties and game speeds.

JovialFeline commented:
That seems reasonable. Funny story, I figured its early appearance would make things _easier_ for two reasons:
- Tanya can be given an attack order for the dog while still airborne.
- The dog is extra sure to be ahead of the incoming soldiers, who might "eat" a click if they mix together.

Figuring anybody would know the first bit is a weird assumption on my part, and my concern with the second is likely overblown with how fast dogs move. It shouldn't be hard to adjust the timing.

Does it sound okay if I keep the dog's appearance the same as a warning but delay the actual attack?

Porenutak commented:
sure

JovialFeline committed:
Add Negotiations mission (Red Alert)

JovialFeline commented:
Delayed the start dog's attack a bit. The timeline (at least as I logged it):
- Tanya activates the trigger at tick 82.
- Tanya touches the ground at 180.
- Dog begins its attack at ~~185~~ 212.

Porenutak reviewed (approved):
LGTM

PunkPun commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/586a7709d6d6ceb87a116e4e88d5871a1d54ec3e...ac29920d1088c5a546f772faca1eb0c663334e36)
